torch/csrc/jit/fuser/compiler.cpp

```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
       std::back_inserter(spec.inputBroadcastGroups()));
 }
 
+// This function moves _grad_sum_to_size nodes along the computation graph
+// of the fusion group to the outputs and then records the shape inputs
+// in order for summation to be applied after the kernel.
+// Note that the correctness relies on the invariant that
+// _grad_sum_to_size is only applied to gradient nodes created by autodiff.
+// This is important because it ensures that in the mul and div nodes only
+// one argument (in the case of diff the numerator) has a summed value.
+// If two arguments to mul had one, we would be in trouble, but thanks
+// to the chain rule, we're OK.
+void processGradSumToSize(KernelSpec& spec) {
+  auto graph = spec.graph();
+
+  // keep track of one sumToSize hitting each output
+  auto& outputMapAndSizes = spec.outputMapAndSizes();
+  JIT_ASSERT(outputMapAndSizes.empty());
+  std::vector<int64_t> outputGradSumToSizes(graph->outputs().size(), -1);
+
+  static OperatorSet commutes_with_SumToSize{{
+      "aten::mul(Tensor self, Tensor other) -> Tensor",
+      "aten::div(Tensor self, Tensor other) -> Tensor",
+      // for div we might check whether we're the first argument
+      "aten::mul(Tensor self, Scalar other) -> Tensor",
+      "aten::div(Tensor self, Scalar other) -> Tensor",
+      "aten::neg(Tensor self) -> Tensor",
+      "aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor",
+      // add this used to be prim::AutogradAdd
+  }};
+  // Scan the graph. We will delete nodes. We want later (in the graph)
+  // _grad_sum_to_size nodes to have priority over earlier ones. Thus
+  // we scan backwards.
+  for (auto it = graph->nodes().rbegin(); it != graph->nodes().rend(); it++) {
+    auto* node = *it;
+    if (node->kind() == aten::_grad_sum_to_size) {
+      use_list uses_to_process(node->output()->uses());
+      while (!uses_to_process.empty()) {
+        auto user = uses_to_process[0].user;
+        auto offset = uses_to_process[0].offset;
+        uses_to_process.erase(uses_to_process.begin());
```

Comment:
Ugh! Please reverse them first and then iterate in order, or this becomes quadratic in the number of uses!!

Reply:
Also, I don't get why would you do that, since uses are not sorted topologically. If that's what you're after, you should sort them explicitly (see the fuser for code that can do that).

Reply:
Ok, reading your code later it seems that you simply want a FIFO queue! In that case, use `std::queue`, which will give you O(1) complexity.

Reply:
`std::queue<Use>` or `std::queue<Use*>`? I don't have any luck with using `std::queue<Use&>`.

Reply:
`std::queue<Use>` is fine. It's a 16B struct.

Reply:
Also, it should be `Use` because `Use*` will get invalidate in case you modify them.

Reply:
Them I'm glad that I picked the right thing. :)